### PR TITLE
Show comparisons between versions in history viewer

### PR DESCRIPTION
**Pull request by NightJar**
Adds the ability to enter "compare mode" when viewing the history for a particular record, allowing a user to select two versions and be presented with a visual comparison between them, according to value by default.

For example `"The Title"` becoming `"Title Story`" would show "The" struck out on a red background (indicating removal), "Title" as normal text, and "Story" on a green background (indicating insertion).

These comparisons are always from the older version to the newer one, chronologically.

Resolves https://github.com/silverstripe/silverstripe-versioned/issues/96, resolves https://github.com/silverstripe/silverstripe-versioned-admin/issues/30, resolves https://github.com/silverstripe/silverstripe-versioned-admin/issues/20 and resolves https://github.com/silverstripe/silverstripe-versioned-admin/issues/17

### CODE REVIEW

File: client/src/components/HistoryViewer/HistoryViewerVersionDetail.js [hunk 1]
```diff
@@ -8,22 +8,90 @@ import { versionType } from 'types/versionType';
...
+   */
+  getListVersions() {
+    const { compare, version } = this.props;
+    if (this.isCompareMode()) {
+      return [compare.versionTo, compare.versionFrom];
+    }
+    return [version];
+  }
+
+  /*
```
Comment (robbieaverill) [inline client/src/components/HistoryViewer/HistoryViewerVersionDetail.js]: Missing the second * - same with `isCompareMode`

File: client/src/components/HistoryViewer/tests/HistoryViewerHeading-test.js [hunk 2]
```diff
@@ -1,32 +1,48 @@
...
 describe('HistoryViewerHeading', () => {
-  it('has four columns when hasActions is true', () => {
-    const component = ReactTestUtils.renderIntoDocument(
-      <HeadingWrapper hasActions />
-    );
+  // Mock select functions to replace the ones provided by mapDispatchToProps
+  const mockOnCompareModeSelect = jest.fn();
+  const mockOnCompareModeUnselect = jest.fn();
+
+  describe('onChange function triggers mapDispatchToProps functions to notify and update the Redux store', () => {
```
Comment (robbieaverill) [inline client/src/components/HistoryViewer/tests/HistoryViewerHeading-test.js]: We should be describing a method e.g. `describe('onChange()'), ...`, not the behaviour - that'd be put in `it('triggers mapDispatch...`

File: client/src/components/HistoryViewer/tests/HistoryViewerVersion-test.js [hunk 3]
```diff
@@ -1,76 +1,222 @@
...
+      Published: false,
+      Publisher: {
+        FirstName: 'Sarah',
+        Surname: 'Smith',
+      },
+      Version: 3,
+    };
+  });
+
+  describe('HistoryViewerVersion', () => {
```
Comment (robbieaverill) [inline client/src/components/HistoryViewer/tests/HistoryViewerVersion-test.js]: We're already describing it at this point. This should be `describe('handleCompare()', ...`

File: client/src/components/HistoryViewer/tests/HistoryViewerVersion-test.js [hunk 4]
```diff
@@ -1,76 +1,222 @@
...
+      const wrapper = shallow(
+        <HistoryViewerVersion
+          version={version}
+          StateComponent={StateComponent}
+          FormActionComponent={FormActionComponent}
+        />
       );
 
-      expect(viewerVersion.textContent).toContain('John Smith');
+      expect(wrapper.instance().getAuthor()).toEqual('John Smith');
```
Comment (robbieaverill) [inline client/src/components/HistoryViewer/tests/HistoryViewerVersion-test.js]: Needs to be within a `describe('getAuthor()', ...`

File: client/src/types/compareType.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { PropTypes } from 'react';
+
+// Describes the data structure for compare version storage
+const compareType = PropTypes.oneOfType([PropTypes.bool, PropTypes.shape({
+  versionFrom: PropTypes.number,
+  versionTo: PropTypes.number
```
Comment (robbieaverill) [inline client/src/types/compareType.js]: Don't we store the version object in the store now?

**Comment by NightJar** [comment]
Seems like this is still pulling the wrong version of `silverstripe/admin` (re: failed `$NPM_TEST`)

File: client/src/components/HistoryViewer/HistoryViewerVersionList.scss [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+.table .history-viewer__row--current {
+  &, &:hover {
+    background-color: $link-color;
+    color: $white;
+  }
+
+  .text-muted {
+    color: $white !important;
```
Comment (maxime-rainville) [inline client/src/components/HistoryViewer/HistoryViewerVersionList.scss]: I understand the root cause of this problem is that `admin` uses `!important` for `.text-muted`. But we're compounding the problem.

Do a separate PR to fix admin.

File: src/Forms/DiffField.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * This form field is both a field object in it's own right, and a decorator for another field type.
+ * It is used to render differences between two data entries into the field type it wraps
+ * e.g. a TextField with the value "Old data" can be trasnformed with {@see DiffTransformation}
+ * and then having setValue called again to load in "New data" - this field will then render
+ * the value "<ins>New</ins> <del>Old</del> data". Most useful in historic version comparisons
+ * {@see SilverStripe\Versioned\Versioned}
+ */
+class DiffField extends HTMLReadonlyField
```
Comment (maxime-rainville) [inline src/Forms/DiffField.php]: This class doesn't appear to have any unit test.

**Comment by robbieaverill** [comment]
I also found another bug:

* Select two versions to compare, where the first is the latest version and the second one is on another page
* Click "Exit" in the compare notice dialog
* The latest version detail form is loaded, notice that the revert button is enabled (should be disabled)

File: client/src/components/HistoryViewer/HistoryViewer.scss [hunk 8]
```diff
@@ -82,6 +60,46 @@
   width: 50%;
 }
 
+.history-viewer__compare-notice {
```
[the remark this answers is not in the document]
Comment (clarkepaul) [reply client/src/components/HistoryViewer/HistoryViewer.scss]: Yeah that should be attached to the top line

File: client/src/components/HistoryViewer/HistoryViewer.scss [hunk 9]
```diff
@@ -18,36 +18,14 @@
...
-  width: 1%;
-  vertical-align: middle;
-}
-
-// Bring any CMS tabs in the detail view closer to the table above it
-.history-viewer__version-detail {
-  .nav-tabs {
-    margin-top: -$panel-padding-y;
-  }
+.history-viewer__compare-button {
```
[the remark this answers is not in the document]
Comment (clarkepaul) [reply client/src/components/HistoryViewer/HistoryViewer.scss]: I think thats a core button style so if we where to fix it we would want to do it globally. They were never meant to have that shadow style but when we updated the Bootstrap version this style started appearing. I'm fine with merging this with type of thing, we should look at it from a pattern library view.

File: client/src/components/HistoryViewer/HistoryViewer.scss (hunk 8 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewer.scss]: I've fixed some left and right margins and can't reproduce this any more, so assume it's resolved

File: client/src/components/HistoryViewer/HistoryViewerVersionDetail.js [hunk 10]
```diff
@@ -113,22 +205,32 @@ HistoryViewerVersionDetail.propTypes = {
...
-  version: versionType.isRequired,
+  version: versionType,
+  compare: PropTypes.oneOfType([
+    PropTypes.shape({
+      versionFrom: versionType,
+      versionTo: versionType,
+    }),
+    PropTypes.bool,
+  ]),
+  previewState: PropTypes.oneOf(['edit', 'preview', 'split']),
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersionDetail.js]: This should really be added into the admin module, it'll also require another webpack config bump. I'll make that PR now but for now I'll comment this saying something like "when webpack-config x.y.z is added, swap this for `import { PreviewStates } from 'state/...`; via admin module"

Sound OK?

File: client/src/components/HistoryViewer/HistoryViewerVersionList.scss (hunk 6 above)
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersionList.scss]: It actually comes from Bootstrap

File: src/Forms/DiffField.php (hunk 7 above)
Comment (robbieaverill) [reply src/Forms/DiffField.php]: Yeah, there's no real logic in this class though. I think Behat test coverage should be sufficient instead - what do you think?

File: client/src/components/HistoryViewer/HistoryViewerVersion.js [hunk 11]
```diff
@@ -32,94 +42,194 @@ class HistoryViewerVersion extends Component {
...
   render() {
     const { version, isActive, StateComponent } = this.props;
 
+    const rowTitle = i18n._t('HistoryViewerVersion.GO_TO_VERSION', 'Go to version {version}');
+
     return (
-      <tr onClick={this.handleClick}>
-        <td>{version.Version}</td>
-        <td>
+      <li className={this.getClassNames()}>
```
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersion.js]: ARIA roles are implicit for lists and list items, adding them throws linting errors:

```
/Users/robbieaverill/dev/ss43/vendor/silverstripe/versioned-admin/client/src/components/HistoryViewer/HistoryViewerVersion.js
  176:7  error  The element li has an implicit role of listitem. Defining this explicitly is redundant and should be avoided  jsx-a11y/no-redundant-roles

/Users/robbieaverill/dev/ss43/vendor/silverstripe/versioned-admin/client/src/components/HistoryViewer/HistoryViewerVersionList.js
  97:9  error  The element ul has an implicit role of list. Defining this explicitly is redundant and should be avoided  jsx-a11y/no-redundant-roles
```

**Comment by robbieaverill** [comment]
Thanks @maxime-rainville, I've pushed a new commit which fixes the margins around the table and uses some variables. I've left some comments on parts that I'm unsure about or don't want to fix =)

**Comment by robbieaverill** [comment]
>I want to raise this separately I think

Actually, I should fix it now since its introduced by this PR

File: client/src/components/HistoryViewer/HistoryViewer.scss (hunk 9 above)
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply client/src/components/HistoryViewer/HistoryViewer.scss]: All good ... I'll let that one go.

File: client/src/components/HistoryViewer/HistoryViewerVersionDetail.js [hunk 12]
```diff
@@ -44,62 +112,86 @@ class HistoryViewerVersionDetail extends PureComponent {
...
-                identifier="HistoryViewer.VersionDetail"
-                schemaUrl={schemaUrl}
-              />
-            </div>
-          </div>
+    const containerClasses = [
+      'flexbox-area-grow',
+      'panel',
+      'panel--scrollable',
+      'panel--padded',
```
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply client/src/components/HistoryViewer/HistoryViewerVersionDetail.js]: This looks like it's fixed.

File: client/src/components/HistoryViewer/HistoryViewerVersionDetail.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply client/src/components/HistoryViewer/HistoryViewerVersionDetail.js]: Sounds good.

File: client/src/components/HistoryViewer/HistoryViewerVersion.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (robbieaverill) [reply client/src/components/HistoryViewer/HistoryViewerVersion.js]: Yeah you're right, will update. It's currently not keyboard accessible either (#20) - will try and kill that with this too

**Comment by robbieaverill** [comment]
>However this acceptance criteria is still not met: "I can share this view by URL to other CMS authors"

We've been subtasking this A/C out of each ticket so far into #1

**Comment by robbieaverill** [comment]
>However this acceptance criteria is still not met: "I can share this view by URL to other CMS authors"

We've been subtasking this A/C out of each ticket so far into #1

**Comment by robbieaverill** [comment]
I've added aria roles to the list and list items, as well as the columns and cells within it, and ensured that it's keyboard accessible (#20).

The Behat tests are failing because the GraphQL queries are no longer running since GraphQL v3 was introduced. I'll need to update this again I think.

**Comment by robbieaverill** [comment]
Ok @maxime-rainville I've investigated this.

* The behat test failure is valid, because a dependency hasn't been merged yet: https://github.com/silverstripe/silverstripe-cms/pull/2221
* The CMS GraphQL queries for graphql v3 have been updated, the corresponding elemental PR is https://github.com/dnadesign/silverstripe-elemental/pull/310

Recapping that permalinked URL A/Cs (from all history viewer epics/stories) so far have been removed and subtasked to #1 for future consideration.

Note that there are 5 PRs for this issue, testing should include them all

**Comment by robbieaverill** [comment]
Wait, I've done bad things with the keyboard accessibility... please hold caller

**Comment by robbieaverill** [comment]
I've updated to switch using an anchor for a span inside list items so that (A) it isn't an anchor with no `href` and (B) using tabindex=0 is justified

**Comment by robbieaverill** [comment]
Thanks for the review @maxime-rainville! I'd like to get Travis green before we merge

**Comment by robbieaverill** [comment]
@raissanorth FYI I've run rebuilt the dist files from your last commit and squashed it, new hash 8611e47

**Comment by robbieaverill** [comment]
I've updated the behat tests.

There's one scenario which is still correctly failing since required dependency https://github.com/silverstripe/silverstripe-cms/pull/2221 hasn't been merged yet.

**Comment by robbieaverill** [comment]
Builds restarted, should be green now
